### PR TITLE
Add luacode and asy as verbatim environments

### LINKED ISSUE #140: The bounder of environment will pass to the injection parser
I want to add parser to luacode environment, and I opened a issue to nvim-treesitter. See [https://github.com/nvim-treesitter/nvim-treesitter/issues/6931?notification_referrer_id=NT_kwDOCJbWB7UxMTU0Mjc2MTIyNzoxNDQxMDI5MTk#issuecomment-2232745541](url). 
So I think it’s good to add a “inner” field to “generic_environment” node, then the user can pass this node to other language parser.

### PR DESCRIPTION
**Pull request by pfoerster**
Adds the following source code environments:
- `asy`
- `luacode`
- `luacode*`

Fixes #140.

### CODE REVIEW

**Comment by clason** [comment]
Generally, it would be enough to just add a single node for (generic) environment contents (that wrap the current list of `word`s); then there is no need to special-case the environments since queries can just inject the language into that node depending on the environment name.

**Comment by clason** [comment]
Also, you forgot to commit the `grammar.json` artefact ;) (This is the most important artefact to commit since it allows generating the parser without node; nvim-treesitter will rely on that in the future.)

**Comment by pfoerster** [comment]
> Generally, it would be enough to just add a single node for (generic) environment contents (that wrap the current list of words); then there is no need to special-case the environments since queries can just inject the language into that node depending on the environment name.

Sure, but one side effect is that the parser will produce garbage nodes under the `environment_content` node when inside a verbatim environment depending on the content (but I assume that the queries will just highlight the entire `environment_content` node instead).

**Comment by clason** [comment]
> Sure, but one side effect is that the parser will produce garbage nodes under the environment_content node when inside a verbatim environment depending on the content (but I assume that the queries will just highlight the entire environment_content node instead).

Hmm, if there's a technical reason, that's different of course. The "generic environment node" is mostly (only) useful for injections. Since the majority of LaTeX environments contain LaTeX, the general grammar should prioritize this situation.

**Comment by clason** [comment]
I apologise if I'm complicating matters; feel free to ignore me.

**Comment by pfoerster** [comment]
> Hmm, if there's a technical reason, that's different of course. 

Yeah, for 

```tex
\begin{lstlisting}[language=bash]
rm -rf "$(biber --cache)"
\end{lstlisting}
```

the generic_environment approach will yield the following parse tree:

```
(source_file
  (biblatex_include
    (curly_group_glob_pattern
      (glob_pattern)))
  (text
    (word)
    (word)
    (generic_command
      (command_name)
      (curly_group
        (text
          (word))))
    (word))
  (text
    (word))
  (ERROR
    (begin
      (curly_group_text
        (text
          (word)))
      (brack_group
        (text
          (word))
        (text
          (word))))
    (text
      (word)
      (operator)
      (word)
      (word))
    (text
      (word)
      (operator)
      (operator)
      (word))
    (text
      (word)
      (generic_command
        (command_name)
        (curly_group
          (text
            (word)))))))
```
I think, the "right" approach is to let the external scanner handle this, when inside a `verbatim` environment.

**Comment by clason** [comment]
Hmm, if the parser will break on such environments, that is of course nothing that can be fixed on a query side. Then, yes, it sounds like there needs to be some way of distinguishing environments with non-LaTeX contents -- which means the parser (or scanner) will still need to maintain a list of such... Then probably nothing is gained by adding a generic environment content node?

**Comment by pfoerster** [comment]
> which means the parser (or scanner) will still need to maintain a list of such

Yes, the parser definitely needs to know which environments contain LaTeX and which environments do not. Otherwise, `tree-sitter` will break with very broken parse trees.

> Then probably nothing is gained by adding a generic environment content node?

For filtering out environments that do not contain LaTeX, then no, it won't help becaue the parser needs to know it anyways.

**Comment by clason** [comment]
That makes sense. Then I recommend continuing with your old approach and just adding the `asy{def}` and `luacode{*}` nodes, similarly to how `pycode` and `sage{block,silent}` are already treated.

Sorry for distracting you!

**Comment by clason** [comment]
> That would be the short-term solution. The long-term solution should handle this in a more generic way within the scanner.

Yes, that sounds good.

File: grammar.js [hunk 1]
```diff
@@ -81,7 +84,9 @@ module.exports = grammar({
         $.verbatim_environment,
         $.listing_environment,
         $.minted_environment,
+        $.asy_environment,
```
Comment (clason) [inline grammar.js:87]: Don't forget `asydef` ;)

**Pull request merged**